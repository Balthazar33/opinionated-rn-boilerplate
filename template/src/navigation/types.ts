import {NavigationProp, ParamListBase, RouteProp, StackActionHelpers, StackNavigationState} from '@react-navigation/native';
import {NativeStackNavigationEventMap, NativeStackNavigationOptions} from 'react-native-screens/lib/typescript/native-stack/types';

export enum StackScreens {
  DummyScreen = 'DummyScreen',
  DummyScreenWithScroll = 'DummyScreenWithScroll',
  ApiCallScreen = 'ApiCallScreen',
  PokemonDetailScreen = 'PokemonDetailScreen',
}

export type NativeStackNavigationProp<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList = string,
  NavigatorID extends string | undefined = undefined
> = NavigationProp<
  ParamList,
  RouteName,
  NavigatorID,
  StackNavigationState<ParamList>,
  NativeStackNavigationOptions,
  NativeStackNavigationEventMap
> &
  StackActionHelpers<ParamList>;

export type NativeStackScreenProps<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList = string,
  NavigatorID extends string | undefined = undefined
> = {
  navigation: NativeStackNavigationProp<ParamList, RouteName, NavigatorID>;
  route: RouteProp<ParamList, RouteName>;
};

export type RootStackParamList = {
  /**
   * Add screens in the following format:
   * <screen name>: <the param(s) it expects>.
   * Specify 'undefined' if the screen does not expect any params.
   */
  DummyScreen: undefined;
  DummyScreenWithScroll: undefined;
  ApiCallScreen: undefined;
  PokemonDetailScreen: {name: string};
  // add more screens here...
};
