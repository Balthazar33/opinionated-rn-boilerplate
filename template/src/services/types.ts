import {AppDispatch} from '@redux/store.utils';

export interface ApiCallWithErrorHandling {
  apiCall: (params: any) => any;
  params: any;
  dispatch: AppDispatch;
}

export interface ApiCallResponse {
  data: null | any;
  error: {
    message?: string;
  };
}

interface Meta {
  headers?: any;
  status: number;
  statusText: string;
}

export interface CommonResponse {
  data: any;
  meta?: {
    response?: Meta | undefined;
  }; // API call meta data
}

export enum FetchErrorStatus {
  PARSING_ERROR = 'PARSING_ERROR',
}

export enum FetchErrorCodes {
  NOT_FOUND = 404,
  UNAUTHORIZED = 401,
  INCORRECT_INPUT = 400,
  SERVER_ERROR = 500,
}

export interface ApiError {
  data?: any;
  originalStatus: FetchErrorCodes | number;
  status: FetchErrorStatus | string;
}

export interface ServerError extends ApiCallWithErrorHandling {
  error: any;
  errorCode?: ErrorType;
}

export enum StatusCodes {
  ERROR400 = 400,
  ERROR401 = 401,
  ERROR404 = 404,
  ERROR500 = 500,
  ERROR300 = 300,
  SUCCESS200 = 200,
  SUCCESS201 = 201,
}

export type ErrorType =
  | StatusCodes.ERROR400
  | StatusCodes.ERROR401
  | StatusCodes.ERROR404
  | StatusCodes.ERROR500;
