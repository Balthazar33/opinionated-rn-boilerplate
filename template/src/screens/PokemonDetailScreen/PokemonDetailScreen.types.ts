import {RootStackParamList, StackScreens} from '@/navigation/types';
import {StackScreenProps} from '@react-navigation/stack';

export type PokemonDetailScreenProps = StackScreenProps<
  RootStackParamList,
  StackScreens.PokemonDetailScreen
>;
