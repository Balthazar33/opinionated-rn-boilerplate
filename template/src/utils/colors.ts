export enum Colors {
  WHITE = '#ffffff',
  BLACK = '#000000',
  PRIMARY_RED = '#ff0000',
  PRIMARY_BLUE = '#0000ff',
  PRIMARY_GREEN = '#00ff00',
  GREY = '#8a8a8a',
  TRANSPARENT = 'transparent',
}

// Opacity range type implementation----------------------
// Ref: https://stackoverflow.com/a/70307091/14431966
type Enumerate<
  N extends number,
  Acc extends number[] = [],
> = Acc['length'] extends N
  ? Acc[number]
  : Enumerate<N, [...Acc, Acc['length']]>;
type Range<F extends number, T extends number> = Exclude<
  Enumerate<T>,
  Enumerate<F>
>;
type OpacityRange = Range<0, 100>;
//--------------------------------------------------------

/**
 * Returns a color with the specified opacity
 * @return color with the specified opacity
 */
export const withOpacity = (color: string, opacity: OpacityRange) => {
  return `${color}${opacity}`;
};
