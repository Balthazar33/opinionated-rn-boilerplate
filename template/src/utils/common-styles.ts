import {StyleSheet} from 'react-native';

export const commonStyles = StyleSheet.create({
  flex1: {
    flex: 1,
  },
  flexGrow1: {
    flexGrow: 1,
  },
  justifyAndAlignCenter: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  rowCenter: {
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  margin0: {
    marginTop: 0,
    marginLeft: 0,
    marginRight: 0,
    marginBottom: 0,
  },
});
