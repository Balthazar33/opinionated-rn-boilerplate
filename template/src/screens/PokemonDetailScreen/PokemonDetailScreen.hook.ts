import {useEffect} from 'react';

import {getDetails} from './PokemonDetailScreen.utils';
import {clearCurrent} from '@/appRedux/slices/pokemonSlice';
import {useAppDispatch, useAppSelector} from '@/appRedux/store.utils';
import {useLazyGetDetailsByNameQuery} from '@/services/testApi/pokemonApi';

export const usePokemonDetailScreen = (name?: string) => {
    const [trigger] = useLazyGetDetailsByNameQuery();
    const {currentPokemon} = useAppSelector(state => state.pokemon) || {};
    const dispatch = useAppDispatch();

    useEffect(() => {
        if (name) {
            getDetails({apiCall: trigger, params: {name}, dispatch});
        }
        return () => {
            dispatch(clearCurrent());
        };
    }, [trigger, dispatch, name]);

    return {currentPokemon};
};
