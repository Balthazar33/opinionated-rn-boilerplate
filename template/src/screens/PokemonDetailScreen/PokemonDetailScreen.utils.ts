import {setPokemonDetails} from '@redux/slices/pokemonSlice';
import {AppDispatch} from '@redux/store.utils';
import {callApi} from '@services/apiCaller';

/**
 * Fetch & set pokemon details
 */
export const getDetails = async ({
  dispatch,
  apiCall,
  params,
}: {
  dispatch: AppDispatch;
  apiCall: (params: any) => any;
  params: any;
}) => {
  try {
    const response = await callApi({
      apiCall,
      params,
      dispatch,
    });
    if (!response.data && !response.isSuccess) {
      // throw error
      throw new Error(response.errorMessage);
    } else {
      // process data
      dispatch(setPokemonDetails(response.data));
    }
  } catch (error) {
    // handle error
    console.log('Error', error);
  }
};
