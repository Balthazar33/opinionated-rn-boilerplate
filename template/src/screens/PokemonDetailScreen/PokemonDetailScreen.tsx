import React, {useEffect} from 'react';
import {StyleSheet, Text} from 'react-native';

import BaseScreen from '../../containers/BaseScreen';
import {sizer} from '../../utils/metrics';
import {SCREEN_PADDING} from '../../utils/constants';
import {useGetDetailsByNameMutation} from '../../services/testApi/pokemonApi';
import {useAppDispatch, useAppSelector} from '../../appRedux/store.utils';
import {getDetails} from './PokemonDetailScreen.utils';
import {clearCurrent} from '../../appRedux/slices/pokemonSlice';
import {PokemonDetailScreenProps} from './PokemonDetailScreen.types';

export const PokemonDetailScreen = ({route}: PokemonDetailScreenProps) => {
  const {name} = route?.params || {};
  const [pokemonDetails] = useGetDetailsByNameMutation();
  const {currentPokemon} = useAppSelector(state => state.pokemon) || {};
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (name) {
      getDetails({apiCall: pokemonDetails, params: {name}, dispatch});
    }

    return () => {
      dispatch(clearCurrent());
    };
  }, [name, dispatch, pokemonDetails]);

  return (
    <BaseScreen style={style.screenStyle}>
      <Text>Name: {name}</Text>
      <Text>Weight: {currentPokemon?.weight}</Text>
    </BaseScreen>
  );
};

const style = StyleSheet.create({
  screenStyle: {
    gap: sizer(16),
    padding: SCREEN_PADDING,
  },
});
