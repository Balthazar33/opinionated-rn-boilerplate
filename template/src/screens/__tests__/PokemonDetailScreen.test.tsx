import React from 'react';

import {it, describe, expect} from '@jest/globals';
import '@testing-library/react-native/extend-expect';

import {PokemonDetailScreen} from '..';
import {renderWithProviders} from '../../utils/test-utils';
import {PokemonDetailScreenProps} from '../PokemonDetailScreen/PokemonDetailScreen.types';

const navigation: Partial<PokemonDetailScreenProps['navigation']> = {
  navigate: jest.fn,
};
const route: Partial<PokemonDetailScreenProps['route']> = {
  params: {name: 'bulbasaur'},
};

describe('PokemonDetailScreen', () => {
  it('renders correctly', () => {
    const elementTree = renderWithProviders(
      <PokemonDetailScreen
        route={route as PokemonDetailScreenProps['route']}
        navigation={navigation as PokemonDetailScreenProps['navigation']}
      />,
    ).toJSON();
    expect(elementTree).toMatchSnapshot();
  });

  it('should display the pokemon name received from the previous screen', () => {
    const {getByText} = renderWithProviders(
      <PokemonDetailScreen
        route={route as PokemonDetailScreenProps['route']}
        navigation={navigation as PokemonDetailScreenProps['navigation']}
      />,
    );
    expect(getByText('bulbasaur', {exact: false})).toBeOnTheScreen();
  });
});
