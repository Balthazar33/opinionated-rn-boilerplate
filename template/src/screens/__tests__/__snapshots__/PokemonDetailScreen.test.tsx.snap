// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonDetailScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
  testID="AppSafeArea"
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        {
          "gap": 32,
          "padding": 32,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#000000",
          "fontFamily": "roboto",
          "fontSize": 20,
        }
      }
    >
      Fetching details for: 
      ivysaur
    </Text>
    <Text
      style={
        {
          "color": "#000000",
          "fontFamily": "roboto",
          "fontSize": 24,
        }
      }
    >
      Name: 
    </Text>
    <Text
      style={
        {
          "color": "#000000",
          "fontFamily": "roboto",
          "fontSize": 24,
        }
      }
    >
      Weight: 
    </Text>
  </View>
</RCTSafeAreaView>
`;
